### ISSUE TITLE
UCGM-T 2Stages training problem

### ISSUE THREAD

**Issue by Peterande** [LINs-lab/UCGM#2]
Hello, excellent work! I would like to ask: Is it necessary to first perform multistep training using the UCGM framework before conducting few - step training? 

Is it feasible to directly use the model obtained through conventional training for fewstep UCGM training?

**Comment by sp12138** [comment]
@Peterande 
Thank you for your interest in our work! Indeed, our framework theoretically fully supports this type of training, though the current code implementation doesn't handle this feature very well.

However, I'm currently working hard to meet the NeurIPS submission deadline. After the DDL passes, I'll provide some implementation examples for your reference. Thanks again for your support!

**Comment by Peterande** [comment]
Glad to hear that! All the best for your great work!

**Comment by Peterande** [comment]
May I ask, if I want to perform fine-tuning from multi-step to few-step while using the classifier free guidance (cfg), should I set the enhanced_ratio to 0? Are there any other hyperparameters that need to be modified?

**Comment by sp12138** [comment]
@Peterande 
Yes, you only need to set `enhanced_ratio` to 0. However, based on my observation, few-step models have very poor support for CFG (this conclusion is the opposite of multi-step models). So while it's theoretically possible to do this, I personally wouldn't recommend it.

**Comment by sp12138** [comment]
@Peterande 
By the way, if you don't need this kind of training enhancement, you can simply set `ema_decay_rate` in the `.yaml` files to 0, which will reduce some GPU memory usage without affecting performance.  

Also, just to add, if you don’t need this training enhancement technique, the existing code implementation already allows direct tuning from multi-step to few-step—regardless of whether the trained multi-step model was obtained from UCGM training or from traditional training methods.

**Comment by Peterande** [comment]
@sp12138 Thank you so much for your response! I applied the fine-tuning you mentioned on some larger models. However, regardless of whether the `enhanced_ratio` is set to 0, noise is generated after several steps (training crashes). There were no observable anomalies from the loss and gradient norm (this might be related to huc). Therefore, I am attempting to use the UCGM's multi-step method to pre-fine-tune the model that was conventionally pre-trained, but it seems that convergence is slow. This could be due to the pre-training time being relatively long compared to the fine-tuning time, and the generated results are still heavily reliant on cfg. Additionally, I noticed the precision for the amp start command is bf16, not fp16. Is bf16 precision sufficient to obtain accurate df_dv_dt?

**Comment by Peterande** [comment]
@sp12138 I have one more question: If the pre-trained model relies on a larger cfg (such as 5.0), should the enhance ratio for UCGM-T's multi-step and few-step methods also be significantly increased?

**Comment by sp12138** [comment]
Hello, @Peterande 
Regarding CFG and enhance_ratio, I can provide a rough conversion formula, namely enhance_ratio=(CFG - 1)/(CFG). For example, for the original model you mentioned, which uses 5.0, the recommended enhance_ratio=(5-1)/5=0.8.
However, please note that the optimal value is usually around this calculated value, which depends on different models and may require manual parameter adjustment.

Generally speaking, it does not take long to tune a pre-trained model into an enhanced version. For the 675M diffusion transformer (pre-trained 1,000,000 steps) I used, it only takes 20,000 steps to complete the tuning process.

**Comment by sp12138** [comment]
Additionally, @Peterande 

Regarding the stability of few-step tuning, I always use 675M transformer and bf16 in my experiments, which is feasible. But since I have not experimented on larger models, I don’t know what impact larger models will have.

In general, I have some suggestions for stabilizing few-step tuning:
(1) Reduce the learning rate or use a more stable optimizer, such as replacing AdamW with RAdam
(2) If your computing resources are sufficient, I suggest you use full-precision training, or use full-precision estimation of df_dv_dt (this may require you to modify the code manually)
(3) You can increase epsilo appropriately, for example, from epsilo=0.005 to epsilo=0.01, but this may have a negative impact on performance
(4) Use a larger scaled_cbl_eps, for example, from scaled_cbl_eps=5 to scaled_cbl_eps=9

Finally, a feasible method is to adopt a strategy of gradually increasing consistc_ratio (from 0.1 to 1.0) to stabilize the tuning process. Or you can try to find the threshold of consistc_ratio first to better locate the problem.

In short, since I don't have enough computing resources to experiment with larger models and higher precision, I may not be able to provide very accurate guidance, but I will try my best to help you. If you have more questions, please feel free to ask!

**Comment by Peterande** [comment]
Thank you for your reply. I have basically tried all the stabilization methods you listed in recent days, but instability still occurs after several steps, and it can only be alleviated. Next, I will try to use 0.8 for multistep training before fewstep training (I used the default 0.44 before). Thank you again for your detailed reply!

**Comment by sp12138** [comment]
Looking forward to your reply. 
@Peterande 
Another very important tip is that if the original multi-step model you use is already fully converged, it may cause extremely serious instability problems. The solution at this time is to use only a fully converged model (e.g. 1000000 steps) as EMA in the trainer, but use a degraded model (ckpt from 500000 or 300000 steps) as the initialization of the model to be trained.
This strategy should be effective, but you may need to carefully check which degraded model is used as the initialization.

**Comment by Peterande** [comment]
```
transport:
  type: Linear
```

@sp12138 It's great to see the update, but why are there no UniGen hyperparameters specified for training in the tuning config? Or did I miss something

**Comment by sp12138** [comment]
@Peterande 
Yes, as mentioned in the [News](https://github.com/LINs-Lab/UCGM?tab=readme-ov-file#fire-news), the code and config to support tuning will be released shortly. We expect to make it available within 1-3 days, as I've just completed the NeurIPS supplementary materials submission and will now focus on reorganizing all the UCGM code.

By the way, I'm curious to ask if the suggestions I provided above could solve the instability issue you mentioned when training the few-step model on larger models.

**Comment by Peterande** [comment]
@sp12138 These methods can alleviate the collapse to some extent, but after a few steps, noise is still generated (early on, there is a trend where the few-step output becomes clearer, but the quality still suffers noticeable degradation). I also tried first fine-tuning a pre-trained model using a multi-step UCGM-T approach, and then performing fewstep training based on its EMA weight, but the collapse still occurred. This might be due to the complex data distribution in larger models, and such collapse also occurs when using sCM. Both codes are evaluated without collapse on imagenet-based small models

**Comment by sp12138** [comment]
Hi, @Peterande 
I've uploaded the tuning code for your reference. I'll also be exploring the instability issue you mentioned by testing it on larger models.
Wishing you the best with your work!!
